### PR TITLE
feat: add genesis functionality

### PR DESCRIPTION
**Pull request by dongsam**
## Description

closes: #66 

## Bullet Points

- [x] Export, Init
- [x] Validation
- [x] test codes

## References

- 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Appropriate labels applied
- [ ] Targeted PR against correct branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: proto/tendermint/farming/v1beta1/genesis.proto [hunk 1]
```diff
@@ -38,31 +61,11 @@ message PlanRecord {
   // plan specifies the plan interface; it can be FixedAmountPlan or RatioPlan
   google.protobuf.Any plan = 1 [(gogoproto.nullable) = false];
```
Comment (hallazzang) [inline proto/tendermint/farming/v1beta1/genesis.proto:62]: Don't we need `(cosmos_proto.accepts_interface) = "PlanI"` option for this field?

File: x/farming/keeper/genesis.go [hunk 2]
```diff
@@ -16,68 +16,66 @@ func (k Keeper) InitGenesis(ctx sdk.Context, genState types.GenesisState) {
...
+		panic(err)
+	}
 
+	for _, record := range genState.PlanRecords {
+		plan, err := types.UnpackPlan(&record.Plan)
+		if err != nil {
+			panic(err)
+		}
+		k.SetPlan(ctx, plan)
+		k.SetGlobalPlanId(ctx, plan.GetId())
```
Comment (hallazzang) [inline x/farming/keeper/genesis.go:30]: What if `PlanRecords` is not sorted correctly? Maybe we should validate that `PlanRecords` is sorted.

File: x/farming/keeper/genesis.go [hunk 3]
```diff
@@ -16,68 +16,66 @@ func (k Keeper) InitGenesis(ctx sdk.Context, genState types.GenesisState) {
...
+			panic(err)
+		}
+		planRecords = append(planRecords, types.PlanRecord{
+			Plan:             *any,
+			FarmingPoolCoins: k.bankKeeper.GetAllBalances(ctx, plan.GetFarmingPoolAddress()),
+		})
+	}
 
-	return types.NewGenesisState(params, planRecords, nil, nil)
+	epochTime, _ := k.GetLastEpochTime(ctx)
```
Comment (hallazzang) [inline x/farming/keeper/genesis.go:60]: `LastEpochTime` may be not present on the store so the second return value `found` can be `false`. In that case, calling `NewGenesisState` with concrete `time.Time` value will not act as expected.

File: x/farming/types/genesis.go [hunk 4]
```diff
@@ -22,6 +27,10 @@ func DefaultGenesisState() *GenesisState {
 		[]PlanRecord{},
 		[]Staking{},
 		[]Reward{},
+		// TODO: zero or nil
+		nil,
```
Comment (hallazzang) [inline x/farming/types/genesis.go:31]: I think we should use `sdk.NewCoins()`(same as `sdk.Coins{}`)

File: x/farming/types/plan.go [hunk 5]
```diff
@@ -279,6 +290,19 @@ func ValidateRatioPlans(i interface{}) error {
 	return nil
 }
 
+// PackPlan converts PlanI to Any
+func PackPlan(plan PlanI) (*codectypes.Any, error) {
+	msg, ok := plan.(proto.Message)
+	if !ok {
+		return nil, fmt.Errorf("cannot proto marshal %T", plan)
+	}
```
Comment (hallazzang) [inline x/farming/types/plan.go:298]: IMO, we don't need this type assertion. `codectypes.NewAnyWithValue` will accept `PlanI` as `proto.Message`.

**Review by hallazzang — CHANGES_REQUESTED** [review]

File: x/farming/keeper/genesis.go (hunk 3 above)
Comment (dongsam) [reply x/farming/keeper/genesis.go:61]: That's right. If there's no value(false), the expectation is to return `time.Time{}` and set it to this empty `time.Time{}` value for export genesis. Can't we do this?

**Pull request merged**